### ISSUE TITLE
Configurable Deprovisioning Process

### ISSUE THREAD

**Issue by njtran** [aws/karpenter#3520]
### Tell us about your request

Currently the deprovisioning process determines nodes to deprovision iteratively for Emptiness, Expiration, Drift, and Consolidation. Once deprovisionable, Karpenter spins up replacement nodes if needed, and then executes graceful termination once that replacement node is ready (cordon, drain, terminate). 

For each of the deprovisioning methods, Karpenter takes an opinion on parts of this process. 
- Karpenter will not deprovision nodes with a `do-not-evict` pod, a blocking PDB.
- Karpenter will wait for a replacement node to be ready before taking action on the deprovisioning node
- Karpenter will only deprovision one node at a time, besides Consolidation.
- Karpenter prioritizes Expiration, then Drift, then Emptiness, then Consolidation
- Karpenter waits to cordon nodes that need deprovisioning until it's going to actually deprovision it. 

Creating this issue to grab some information from the community on what parts of the deprovisioning process doesn't align with their assumptions. 

Some related issues: aws/karpenter-core#623 aws/karpenter-core#622 aws/karpenter#3350 

### Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?

Mostly just an RFC here. 

### Are you currently working around this issue?

No need.

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by runningman84** [comment]
Will this also take care of the problem that the current drift detection does not wait for all pods to be ready? The current implementation causes outages if you have pods which need 10 minutes to be ready.

**Comment by runningman84** [comment]
I think we do not have a PDB for everything. The problem is that if you have a huge cluster with different teams it is difficult to ensure that each and every application has a corresponding PDB...

**Comment by tzneal** [comment]
You might look at something like https://kyverno.io/policies/other/require_pdb/require_pdb/ to enforce PDBs. They are the solution for the problem you face. We need some indication of how many pods can be terminated at once, and a PDB is the method of specifying that.  We can't know otherwise.

**Comment by wmgroot** [comment]
PDBs are the way to specify if your pods have disruption requirements. The Kubernetes node drain process respects PDBs. Writing any additional logic around pod health other than the existing drain process would be duplicating the K8s logic that already exists.

**Comment by wmgroot** [comment]
Was a fully capable node de-provisioning lifecycle that supports all bullet points listed in the original issue implemented? If so, Is there an associated PR or documentation describing what's supported?

We're very interested in the progress of this issue as a major blocker for adoption of Karpenter by our organization and are unsure what the current status of "closed" means for this issue.

**Comment by njtran** [comment]
Thanks for messaging here. @wmgroot, I had opened this as an RFC to query the community and see which parts of the process users may want to configure. I'll re-open. Can you share which bullet points are major blockers? What of the current process is not what you desire?

**Comment by wmgroot** [comment]
Thanks for the clarification. These are the points that we're most interested in before considering using Karpenter for our production clusters.

Our thoughts on each point.
**1. Karpenter will not deprovision nodes with a do-not-evict pod, a blocking PDB.**

1a. We do not require support for "do-not-evict".
1b. We're fine if Karpenter begins to drain nodes as part of the deprovisioning process without checking to see if a pod will be blocked due to a PDB, because the existing drain process already handles waiting for PDBs. 
1c. We would like for there to be a configurable "drain timeout" after which the node is forcibly drained, without regard for PDBs or other blocking mechanisms. This allows us to provide our users with a window of time to gracefully terminate their pods, but still ensures that we can apply node level security updates by replacing nodes in an automated way. The PR adding support for this feature in ClusterAPI can be found [here](https://github.com/kubernetes-sigs/cluster-api/pull/3662).

**2. Karpenter will wait for a replacement node to be ready before taking action on the deprovisioning node.**

2a. We think this is necessary. We've run into edge cases with Cluster Autoscaler where fewer nodes than necessary were running to support all running pods, which causes pain for our users and wastes engineering time to investigate. These were generally caused by bugs and misaligned assumptions in interactions between CAPI and Cluster Autoscaler which are difficult to solve due to the implementation history of CA, and we'd like to avoid the same problems entirely with a switch to Karpenter.
2b. Having the option to deprovision before replacement could be useful to optimize for situations where there's not enough availability of a particular instance type in an AZ. CAPI supports this through the use of maxSurge (provision then deprovision) vs maxUnavailable (deprovision then provision). Both options are independent and can be used in conjunction for maximum flexibility during the node lifecycle. This is not a requirement for production use, just a nice to have.

**3. Karpenter will only deprovision one node at a time, besides Consolidation.**

3a. Cycling one node at a time is not a requirement for production.
3b. The ability to control how many nodes are cycled simultaneously is a hard requirement for production. If our drain timeout (1c) is set to 24 hours and we have 100 nodes in the cluster, it could take 100 days in the worst case for all nodes to be cycled, which would be unacceptable given that we have requirements to cycle nodes for security updates within 30 days.

**4. Karpenter prioritizes Expiration, then Drift, then Emptiness, then Consolidation.**

This overall statement seems reasonable to us.

**5. Karpenter waits to cordon nodes that need deprovisioning until it's going to actually deprovision it.**

This also seems reasonable to us, and a good requirement to ensure we minimize the time where nodes are unschedulable during rotation.


Thanks for reaching out for clarification, we're happy to provide more details about our use cases if you'd find more information useful.

**Comment by njtran** [comment]
Sorry for the delay,
To your points:

1c. Drain timeout is being tracked here https://github.com/aws/karpenter/issues/2481
2a. This is currently done, we won't continue with a disruption action for a current node if the pods won't reschedule.
2b. Makes sense, one option here could be to add configuration surface to modify the replacement logic to just wait for the node launch to become successful, which users could enable if they assume that the bootstrapping/initialization logic would succeed, or can tolerate if it doesn't. This would enable us to modify the logic to do a MaxSurge like you're describing. This would be best tracked through this: https://github.com/aws/karpenter/issues/3350
3b. This will be fixed once the work detailed in https://github.com/aws/karpenter-core/pull/516, specifically in https://github.com/aws/karpenter/issues/3520.
5. Is now being tracked with https://github.com/aws/karpenter-core/issues/624. 

You should look at https://github.com/aws/karpenter/issues/1738 and see if any of the linked issues are also important to you.

Since it looks like all your questions are answered and are being tracked elsewhere, I'm going to close this issue. Please cut another issue if you have another use-case we haven't discussed that you can't find a github issue for!
